Sum-check protocol

Al-Kindi-0 opened the pull request:
## Describe your changes
Implements the sum-check protocol for use in the virtual bus as in #1182 .

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

plafer commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        if let Some(openings_claim) = openings_claim {
+            if openings_claim.evaluation_point != evaluation_point {
+                return Err(Error::WrongOpeningPoint);
+            }
```
Is this check necessary? The prover can just put anything in `openings_claim.evaluation_point`, unrelated to the openings. So IIUC, we could just remove the field `OpeningClaims::evaluation_point`.

Al-Kindi-0 replied:
You are right but there are two reasons for having it as part of the claim:

1. In the case where we have many different runs of different sum-checks, we would have different opening points against some of the same oracles. At this point, one can usually do a batch opening argument in order to reduce the cost of the opening proof. This is probably not immediately relevant for us but could be in the future.
2. The more relevant reason for us is so that the STARK prover would have all the data necessary in order to build the auxiliary trace helper columns in order to do the (multi-linear) opening proof.

To summarize, we could, and probably, should remove the evaluation points (in order to reduce the size of the GKR proofs), but having it this way is convenient at the moment and potentially necessary in the future.

plafer commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+            } else {
+                Ok(openings_claim)
+            }
+        } else {
+            Err(Error::NoOpeningsProvided)
+        }
```
If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. *not* an `Option`)

plafer replied:
I just noticed the comment on `Proof`:

```rs
/// Openings is an [Option] as there are situations where we would like to run the sum-check
/// protocol for a certain number of rounds that is less than the number of variables of the
/// multi-linears. This is the case for example when we have a merged polynomial.
```

What is meant by a "merged polynomial"?

Al-Kindi-0 replied:
> If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. _not_ an `Option`)

I have changed this in the subsequent PR as this was a poor design choice

Al-Kindi-0 replied:
> I just noticed the comment on `Proof`:
> 
> ```rust
> /// Openings is an [Option] as there are situations where we would like to run the sum-check
> /// protocol for a certain number of rounds that is less than the number of variables of the
> /// multi-linears. This is the case for example when we have a merged polynomial.
> ```
> 
> What is meant by a "merged polynomial"?

Fixed now

plafer reviewed (commented):
Looks good :) The documentation is very well done!

Not a full review yet

bitwalker commented on processor/Cargo.toml:
```diff
@@ -23,9 +23,10 @@ internals = ["miden-air/internals"]
 std = ["vm-core/std", "winter-prover/std"]
 
 [dependencies]
+thiserror = { version = "1.0", default-features = false }
```
Once #1277 is merged, we will have a version of `thiserror` (and `miette`) that work in no-std environments, using forks that I'm maintaining. Once `error_in_core` is stabilized, those forks can go away and we can use the mainline crates, but at least we'll have proper error infrastructure to lean on in the meantime

Al-Kindi-0 replied:
> Once #1277 is merged, we will have a version of `thiserror` (and `miette`) that work in no-std environments, using forks that I'm maintaining. Once `error_in_core` is stabilized, those forks can go away and we can use the mainline crates, but at least we'll have proper error infrastructure to lean on in the meantime

Thank you Paul

plafer commented on processor/src/trace/virtual_bus/sum_check/domain.rs:
```diff
@@ -0,0 +1,65 @@
+/// Computes the barycentric weights for a set of interpolation points.
+pub fn barycentric_weights<E: FieldElement>(points: &[E]) -> Vec<E> {
+    let n = points.len();
+    let tmp = (0..n)
```
nit: would rename to `weight_denominators` or something of the like

Al-Kindi-0 replied:
changed

plafer replied:
Actually I was referring to the `tmp` variable to be renamed to `weight_denominators`, since it contains the denominators before they get inverted. I think `barycentric_weights()` is a good name for the function, since it agrees with the definition of barycentric weights [in the literature](https://people.maths.ox.ac.uk/trefethen/barycentric.pdf).

plafer commented on processor/src/trace/virtual_bus/sum_check/domain.rs:
```diff
@@ -0,0 +1,65 @@
+
+/// Computes the value of the polynomial with minimal degree interpolating the set `{(x_i, y_i)}`
+/// at the point `r` given pre-computed barycentric weights.
+pub fn evaluate_barycentric<E: FieldElement>(
```
I would either make `barycentric_weights()` and `evaluate_barycentric()` private, or put them in some `barycentric` utils module - this module should be only for sum-check domain.

For now though, I'd probably just make them private, as using `EvaluationDomain` will always be cleaner than using these 2 explicitly. And I'd also just inline `evaluate_barycentric()`

plafer commented on processor/src/trace/virtual_bus/sum_check/domain.rs:
```diff
@@ -0,0 +1,65 @@
+/// Computes the value of the polynomial with minimal degree interpolating the set `{(x_i, y_i)}`
+/// at the point `r` given pre-computed barycentric weights.
+pub fn evaluate_barycentric<E: FieldElement>(
+    x_i: &[E],
+    y_i: &[E],
```
nit: I would rename to `xs` and `ys`, since `x_i` usually stands for "the ith `x`", and it sort of clashes with the iteration variable `i` in `x_i[i]` (and similarly for `y_i`)

plafer commented on processor/src/trace/virtual_bus/sum_check/domain.rs:
```diff
@@ -0,0 +1,65 @@
+
+    let l_x: E = x_i.iter().fold(E::ONE, |acc, &x_i| acc * (r - x_i));
+
+    let sum = (0..x_i.len()).fold(E::ZERO, |acc, i| {
+        let w_i = barycentric_weights[i];
+        acc + (w_i / (r - x_i[i]) * y_i[i])
```
```suggestion
        let term = (w_i / (r - x_i[i])) * y_i[i];
        acc + term
```

nit: I would add the extra parentheses so as to not rely on precedence rules which are easy to forget/get wrong. But this does become a bit paren-heavy, so I split it over 2 lines (feel free to re-arrange)

plafer replied:
Is it worth using `batch_inversion()` here on the term `(r - x_i[i])` so as to avoid dividing `n` times?

plafer commented on processor/src/trace/virtual_bus/sum_check/mod.rs:
```diff
@@ -0,0 +1,277 @@
+        assert!(result.is_err())
+    }
+
+    struct PlainQueryBuilder;
```
```suggestion
    struct ProjectionQueryBuilder;
```

Since this is the case where `f_i` returns the `i`th input

plafer commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+/// the Verifier can do by herself. For example, this is the case for periodic columns where given
+/// `(r_0, ... ,r_{\nu - 1})` the Verifier can evaluate `f_i(r_0, ... ,r_{\nu - 1})` un-assisted.
+///
+/// In the case where there are no `f_i(r_0, ... ,r_{\nu - 1})` which can be computed by the
```
This is a bit misleading - we always have `f_i`s. The case being described here is the case where `f_i` returns the `i`th input. Saying that there are no `f_i` contradicts the initial problem statement, and is a bit confusing. I would just leave this out

Al-Kindi-0 replied:
Removed

plafer commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+
+/// Contains the logic for building the final query made to the virtual polynomial.
+///
+/// During the last step of the sum-check protocol, the Verifier must evaluate the virtual
+/// polynomial at a random point `(r_0, ... ,r_{\nu - 1})`. To do this, the Verifier asks the Prover
```
The term "virtual polynomial" doesn't appear in the `SumCheckProver` docstring that we use as central point of documentation of the algorithm. Could we clearly document what is meant by "virtual polynomial" there?

Al-Kindi-0 replied:
Updated

plafer commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+/// the current evaluation at `x` in {2, ... , d_max}.
+fn sumcheck_round<E: FieldElement>(
+    polynomial: &dyn CompositionPolynomial<E>,
+    mls: &mut [MultiLinear<E>],
```
```suggestion
    mls: &[MultiLinear<E>],
```

plafer commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+
+        // run the first round of the protocol
+        let mut round_proof = sumcheck_round(&self.composition_poly, mls);
+        round_proofs.push(round_proof);
+        // reseed with the s_0 polynomial
+        coin.reseed(H::hash_elements(&round_proofs[0].poly_evals));
```
nit: I would first reseed the coin using `&round_proof.poly_evals`, and then push `round_proof` in `round_proofs`

Al-Kindi-0 replied:
Makes sense

plafer commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+    pub fn prove(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinear<E>],
```
We have a name for `g()`: the "composition polynomial". Could we also have a name for the `f_i`s, that's more descriptive than "multilinears"?

Al-Kindi-0 replied:
We could use: multi-linear extension polynomials but I am not sure this is more descriptive

plafer commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+            current_round_claim = new_round_claim;
+
+            // reseed with the s_i polynomial
+            coin.reseed(H::hash_elements(&round_proofs[i].poly_evals));
```
Similarly, I would reseed the coin on `round_proof`

plafer commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+            *res = self.evaluations[i << 1]
+                + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+        }
+        *self =
+            Self::from_values(&result).expect("should not fail given that it is a multi-linear");
```
`bind_assign(&mut self)` to me suggests that this is done in-place, and that there is a performance benefit over using `bind()`. Instead, this is a new function for `*self = self.bind(...)`. I would remove this function and use `bind()` everywhere instead

Al-Kindi-0 replied:
I would like it to be in place but not sure about the most straightforward way to do this
I removed and renamed as suggested

plafer commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+            );
+
+            // fold each multi-linear using the round challenge
+            mls.iter_mut().for_each(|ml| ml.bind_assign(round_challenge));
```
Question: are there benefits to binding on each iteration, as opposed to keeping the initial multilinears as-is, and prepending the previous rounds challenges `[r0, ..., rm]` to the input? I'm not sure of the relative costs of binding vs running `sumcheck_round` on more variables.

Al-Kindi-0 replied:
I am not sure about the meaning of prepending the previous round challenges

plafer replied:
Basically, we'd leave the multilinears untouched, and call them e.g. as `ml.evaluate([r0, ..., rm, x])`, where `r0, ... rm` are the previous rounds' claims.

Basically, run the algorithm as described in the literature - keeping the `r_i`s around, and no binding.

Al-Kindi-0 replied:
Ok, and what are we comparing here exactly?

> I'm not sure of the relative costs of binding vs running sumcheck_round on more variables.

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+        .fold(E::ZERO, |acc, (x_i, y_i)| acc + *x_i * *y_i)
+}
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
```
I think this is used only in few places, so it shouldn't be that taxing in my opinion

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+
+    let mut evals: Vec<E> = vec![E::ONE; n];
+    let mut size = 1;
+    for query in query.iter() {
```
Changed to `r_i`  in the body and left `query` as it is more descriptive

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+
+impl<E: FieldElement> MultiLinear<E> {
+    /// Constructs a [MultiLinear] from an array of values in the field.
+    pub fn new(values: Vec<E>) -> Result<Self, Error> {
```
Agreed and changed

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Constructs a [MultiLinear] from a slice of values in the field.
+    pub fn from_values(values: &[E]) -> Result<Self, Error> {
```
Done

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+
+/// The evaluations of a mult-linear polynomial over the boolean hyper-cube {0 , 1}^ν.
+#[derive(Clone, Debug)]
+pub struct MultiLinear<E: FieldElement> {
```
Changed

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
+pub fn tensorize<E: FieldElement>(query: &[E]) -> Vec<E> {
```
Will do that in the next PR. 
Will leave this conversation here so that I remember

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+/// protocol prover.
+/// The sum-check protocol is an interactive protocol (IP) for proving the following relation:
+///
+/// v = \sum_{(x_0,\cdots, x_{\nu - 1}) \in \{0 , 1\}^{2^{\nu}}}
```
Good catch

Al-Kindi-0 commented on processor/src/trace/virtual_bus/multilinear/mod.rs:
```diff
@@ -0,0 +1,151 @@
+/// A multi-variate polynomial for composing individual multi-linear polynomials.
+pub trait CompositionPolynomial<E: FieldElement>: Sync + Send {
+    /// The number of variables when interpreted as a multi-variate polynomial.
+    fn num_variables(&self) -> usize;
+
+    /// Maximum degree in all variables.
+    fn max_degree(&self) -> usize;
```
Changed now

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+            round_proofs,
+        } = proof;
+
+        let mut claimed_evaluation = claim;
```
Changed

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+///
+/// In the case where there are no `f_i(r_0, ... ,r_{\nu - 1})` which can be computed by the
+/// Verifier alone, the output of [Self::build_query] will be just the provided openings.
+pub trait FinalQueryBuilder {
```
Changed

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+/// In the case where there are no `f_i(r_0, ... ,r_{\nu - 1})` which can be computed by the
+/// Verifier alone, the output of [Self::build_query] will be just the provided openings.
+pub trait FinalQueryBuilder {
+    type Field: FieldElement;
+
+    fn build_query(
```
Changed

plafer replied:
The current change made `CompositionPolyQueryBuilder` generic over `E`. I think it would be better if `build_query()` was generic over `E` instead.

This is because `impl<E> CompositionPolyQueryBuilder<E> for X` results in multiple trait implementations (one per `E`). Hence you can no longer do `X.build_query()`, because the compiler no longer knows which `build_query()` you're referring to (except in [cases](https://github.com/0xPolygonMiden/miden-vm/blob/9764c43712fcb6f5b7a25d973c321116c992ea53/processor/src/trace/virtual_bus/sum_check/verifier/mod.rs#L101) where the implementation is also generic over that implementation). It would also remove the need for [this `PhantomData`](https://github.com/0xPolygonMiden/miden-vm/blob/9764c43712fcb6f5b7a25d973c321116c992ea53/processor/src/trace/virtual_bus/sum_check/mod.rs#L226).

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:
```diff
@@ -0,0 +1,152 @@
+/// `r_i` is the randomness sent by the Verifier at each round.
+pub struct SumCheckVerifier<B, E, P, C, H, V>
+where
+    B: StarkField,
```
Changed

Al-Kindi-0 commented on processor/src/trace/virtual_bus/sum_check/prover/mod.rs:
```diff
@@ -0,0 +1,289 @@
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    C: RandomCoin<Hasher = H, BaseField = B>,
```
But then would have the same issue with `H`?

plafer replied:
I'm no longer sure why I suggested this, looks fine to me as is.

Al-Kindi-0 commented:
I have addressed most of the feedback but for the ones related to the evaluation domain. The reason for this is that we will be moving to the coefficient in a subsequent PR and we can implement most of the feedback as part of that PR.

plafer commented:
Should we just close in PR in favor of #1307?

Al-Kindi-0 commented:
The implementation has now moved to using the coefficient form for the round polynomial instead of the evaluation one. With this change, we avoid having to deal with evaluation domains.
Also, regarding `CompositionPolyQueryBuilder`, I had to revert back to the previous implementation which had the trait generic over the field.  The reason been that for the following typical use case, the previous approach didn't work as (maybe there is a workaround I am not aware of)
```rust
/// A [`FinalQueryBuilder`] for the sum-check verifier used for the final sum-check.
#[derive(Default)]
struct GkrMergeQueryBuilder<E> {
    gkr_eval_point: Vec<E>,
    merge_rand: Vec<E>,
}

impl<E: FieldElement> CompositionPolyQueryBuilder<E> for GkrMergeQueryBuilder<E> {
    fn build_query(&self, openings_claim: &FinalOpeningClaim<E>, evaluation_point: &[E]) -> Vec<E> {
        let eq_at_gkr_eval_point = EqFunction::new(self.gkr_eval_point.clone());
        let mut rand_sumcheck = self.merge_rand.clone();
        rand_sumcheck.extend_from_slice(evaluation_point);
        let eq = eq_at_gkr_eval_point.evaluate(&rand_sumcheck);
        let mut query = openings_claim.openings.clone();
        query.push(eq);
        query
    }
}
```

As for closing this PR, I agree. I think that there no outstanding comments left here. The review for the change to coefficient form could be done in the other PR.

plafer commented:
Closing; let's move all discussion to #1307

plafer commented:
Reopening - @Al-Kindi-0 could you merge your recent commits into #1307, and then close this one?

Al-Kindi-0 commented:
> Reopening - @Al-Kindi-0 could you merge your recent commits into #1307, and then close this one?

Done! I will now close this.
